### PR TITLE
issue 217 mongodb testcontainers

### LINKED ISSUE #217: Use TestContainers instead of flapdoodle's embedded Mongo

### LINKED ISSUE #588: Dependency has wrong scope - testRuntimeOnly instead of testImplementation
When creating a project via micronaut starter with testcontainers and mariadb features it creates following
gradle dependencies.

```groovy
dependencies {
    implementation("io.micronaut:micronaut-validation")
    implementation("io.micronaut.groovy:micronaut-runtime-groovy")
    implementation("javax.annotation:javax.annotation-api")
    implementation("io.micronaut:micronaut-http-client")
    runtimeOnly("ch.qos.logback:logback-classic")
    runtimeOnly("org.mariadb.jdbc:mariadb-java-client")
    testRuntimeOnly("org.testcontainers:mariadb")
}
```
`testRuntimeOnly("org.testcontainers:mariadb")` should be `testImplementation("org.testcontainers:mariadb")`

### PR DESCRIPTION
**Pull request by wetted**
- changing mongodb feature to use testcontainers instead of flapdoodle
- adding testcontainer annotations dependencies spock/junit5 to testcontainers feature

Closes #217 
Closes #588

### CODE REVIEW

**Review by alvarosanchez — APPROVED** [review]

File: starter-core/src/main/java/io/micronaut/starter/feature/build/gradle/templates/dependencies.rocker.raw [hunk 1]
```diff
@@ -492,13 +492,13 @@ dependencies {
...
-@if (features.contains("mongo-reactive") || features.contains("mongo-sync")) {
-    @dependency.template("de.flapdoodle.embed", "de.flapdoodle.embed.mongo", "testImplementation", "2.0.1")
-}
 @if (features.contains("neo4j-bolt")) {
     @dependency.template("org.neo4j.test", "neo4j-harness", "testRuntimeOnly", null)
 }
 @if (features.contains("testcontainers")) {
+@if (features.contains("mongo-reactive") || features.contains("mongo-sync")) {
+    @dependency.template("org.testcontainers", "mongodb", "testImplementation", null)
+}
```
Comment (ilopmar) [inline starter-core/src/main/java/io/micronaut/starter/feature/build/gradle/templates/dependencies.rocker.raw:501]: Question: if `mongo` is now nested inside `testcontainers`, is it possible to select only mongo without selected testcontainers?

File: starter-core/src/test/groovy/io/micronaut/starter/feature/database/MongoGormSpec.groovy [hunk 2]
```diff
@@ -27,7 +45,19 @@ class MongoGormSpec extends BeanContextSpec {
         then:
         template.contains('implementation("io.micronaut.groovy:micronaut-mongo-gorm")')
         template.contains('implementation("io.micronaut.mongodb:micronaut-mongo-reactive")')
-        template.contains("testImplementation(\"de.flapdoodle.embed:de.flapdoodle.embed.mongo:2.0.1\")")
+    }
+
+    void "test testcopntainers dependencies are present for gradle"() {
```
Comment (ilopmar) [inline starter-core/src/test/groovy/io/micronaut/starter/feature/database/MongoGormSpec.groovy:50]: type in `testcopntainers`

Comment (alvarosanchez) [reply starter-core/src/test/groovy/io/micronaut/starter/feature/database/MongoGormSpec.groovy:50]: > type in `testcopntainers`

Typo in `type` :trollface:

Comment (ilopmar) [reply starter-core/src/test/groovy/io/micronaut/starter/feature/database/MongoGormSpec.groovy:50]: hahaha :rofl:

File: starter-core/src/main/java/io/micronaut/starter/feature/build/gradle/templates/dependencies.rocker.raw (hunk 1 above)
Comment (wetted) [reply starter-core/src/main/java/io/micronaut/starter/feature/build/gradle/templates/dependencies.rocker.raw:501]: No, if a mongodb feature is selected, you get the `testcontainers` dependencies for it. It works like that already for the other databases that use `testcontainers` (MySql, Oracle, MariaDb, etc). For each of them, if you select the database feature, but not the `testcontainers` feature, then `testcontainers` dependencies (and now junit/spock annotations for it) are included automatically. 

OTOH, if the `testcontainers` feature is selected, then the `application-test.yml` file with the "tc" driver url is generated. However, mongodb `testcontainers` doesn't use the special "tc" driver url, so it ends up making no difference whether one selects `testcontainers` or not. Given that, the logic is consistent for selecting just the database feature for mongodb as it is for the others.

Maybe `testcontainers` should just be an automatic feature for the databases?

Comment (ilopmar) [reply starter-core/src/main/java/io/micronaut/starter/feature/build/gradle/templates/dependencies.rocker.raw:501]: Sounds good. Thanks for the clarification.

**Review by ilopmar — APPROVED** [review]

**Review by ilopmar — APPROVED** [review]

**Review by alvarosanchez — APPROVED** [review]

**Pull request merged**